gmeta/codegen/src/lib.rs

```diff
@@ -1,518 +1,503 @@
-
-    let state_type = type_item.default.expect("Checked above").1;
-
-    Ok(state_type)
+fn has_attributes(
+    attributes: impl IntoIterator<Item = impl Borrow<Attribute>>,
+    message: impl Display,
+) -> Result<(), TokenStream> {
+    let mut attributes = attributes.into_iter();
+
+    if let Some(attribute) = attributes.next() {
+        let span = attributes.fold(attribute.borrow().span(), |span, attribute| {
+            span.join(attribute.borrow().span()).unwrap_or(span)
+        });
+
+        Err(error(span, message))
+    } else {
+        Ok(())
+    }
 }
 
-fn parse_fn_item(item: TraitItem, state: &syn::Type) -> Result<ItemFn, TokenStream> {
-    // TODO: validate here.
-    let mut trait_item_method = match item {
-        TraitItem::Method(trait_item_method) => trait_item_method,
-        _ => {
-            return Err(compile_error(
-                item,
-                "Unsupported trait item: it should contain 1 type and fns",
-            ))
+/// Generates metawasm functions.
+///
+/// An example of the expected structure:
+/// ```
+/// use gstd::prelude::*;
+///
+/// #[derive(Decode, Encode, TypeInfo)]
+/// pub struct StateType;
+///
+/// #[derive(Encode, TypeInfo)]
+/// pub struct SomeReturnType;
+///
+/// #[derive(Decode, TypeInfo)]
+/// pub struct SomeArg;
+///
+/// #[gmeta::metawasm]
+/// pub mod metafuncs {
+///     pub type State = StateType;
+///
+///     /// Documentation...
+///     pub fn some_function(_: State) -> SomeReturnType {
+///         unimplemented!()
+///     }
+///
+///     pub fn another_function_but_with_arg(mut _state: State, _arg: SomeArg) -> State {
+///         unimplemented!()
+///     }
+///
+///     /// DOCSdocsDoCsdOcSDoCS...
+///     pub fn function_with_multiple_args(
+///         _state: State,
+///         mut _arg1: SomeArg,
+///         _arg2: u16,
+///         mut _arg3: u32,
+///     ) -> SomeReturnType {
+///         unimplemented!()
+///     }
+/// }
+/// # fn main() {}
+/// ```
+///
+/// # Syntax
+///
+/// - This attribute **must** be used on the `pub`lic `mod` container with the
+/// `metafuncs` identifier.
+/// - The first item in the module **must** be a `pub`lic `type` alias with the
+/// `State` identifier. The type for which `State` will be an alias **must**
+/// implement [`Decode`] trait.
+///
+/// Usually the state type should be imported from the implemented associative
+/// [`Metadata::State`](../gmeta/trait.Metadata.html#associatedtype.State) type
+/// from the contract's `io` crate.
+///
+/// - The rest of items **must** be `pub`lic functions.
+/// - The first argument's type of metafunctions **must** be `State`.
+///
+/// In addition to the mandatory first argument, functions can have additional
+/// ones.
+///
+/// - The maximum amount of additional arguments is 18 due restrictions of the
+/// SCALE codec.
+/// - All additional arguments **must** implement the [`Decode`] &
+/// [`TypeInfo`] traits.
+/// - A function **mustn't** return `()` or nothing.
```

Comment:
```suggestion
/// - A function **mustn't** return `()`.
```

If nothing specified, it means compiler places `-> ()` in Rust

Reply:
You're right. But I think this can be left for Rust newbies. And furthermore, there's a difference between no return type and the unit type for the macro. That's why it has different error messages for both cases.

Reply:
Ok, let it be